In [1]:
import pandas as pd
import numpy as np
import os
from datetime import datetime

In [2]:
path='Files used in current analysis/Linkedin_Job_Details_26-Sep-2023-to-26-Nov-2023_15875-jobs_685-pages.csv'

In [3]:
df=pd.read_csv(path)

In [4]:
df

Job Title  \
0                              LinkedIn Data Extractor   
1                                 It Management Intern   
2                               SQL Database Developer   
3                      Business Intelligence Developer   
4                           Web Developer - Internship   
...                                                ...   
15870  Manager – Growth Chat & SMS (Digital Marketing)   
15871                         Release Support Engineer   
15872                  Information Security Specialist   
15873        FUNCTIONAL CONSULTANT ORACLE FUSION - HCM   
15874                              Community Associate   

                                                Job Info  \
0            EcomFleet · Pakistan (Remote)  \n1 week ago   
1      Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2      Murkez Technologies · Lahore, Punjab, Pakistan...   
3      DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4      ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                  ...   
15870  Bank Alfalah Limited · Karāchi, Sindh, Pakista...   
15871           Turing · Pakistan Reposted  \n1 week ago   
15872  DPL · Islamabad, Islāmābād, Pakistan  \n1 week...   
15873  Level 3 BOS · Karachi Division, Sindh, Pakista...   
15874  The Hive · Karachi Division, Sindh, Pakistan  ...   

                                             Job Type  \
0                                           Full-time   
1                                           Full-time   
2                                           Full-time   
3                                           Full-time   
4                                           Full-time   
...                                               ...   
15870  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15871        Remote\n\n \nFull-time\n\n \nEntry level   
15872       On-site\n\n \nFull-time\n\n \nEntry level   
15873  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15874                         On-site\n\n \nFull-time   

                                            Job Link  \
0      https://www.linkedin.com/jobs/view/3727054368   
1      https://www.linkedin.com/jobs/view/3723395858   
2      https://www.linkedin.com/jobs/view/3727034665   
3      https://www.linkedin.com/jobs/view/3727029273   
4      https://www.linkedin.com/jobs/view/3727034696   
...                                              ...   
15870  https://www.linkedin.com/jobs/view/3767355025   
15871  https://www.linkedin.com/jobs/view/3766809535   
15872  https://www.linkedin.com/jobs/view/3772530977   
15873  https://www.linkedin.com/jobs/view/3771943212   
15874  https://www.linkedin.com/jobs/view/3767944844   

                                               employees  
0                                        11-50 employees  
1                                        11-50 employees  
2                                       51-200 employees  
3                                        11-50 employees  
4                    Mobeen Naqvi is hiring for this job  
...                                                  ...  
15870                   5,001-10,000 employees · Banking  
15871  1,001-5,000 employees · IT Services and IT Con...  
15872   51-200 employees · IT Services and IT Consulting  
15873  201-500 employees · Outsourcing and Offshoring...  
15874                                    11-50 employees  

[15875 rows x 5 columns]

### checking duplicates

In [5]:
len(df)

15875

In [6]:
df['Job ID']=df['Job Link'].apply(lambda x:int(x.split('/')[5]))

In [7]:
len(df['Job ID'].unique())

12010

In [8]:
len(df)-len(df['Job ID'].unique())

3865

In [1823]:
df['Job ID'].value_counts()

Job ID
3720211981    13
3707088816    11
3733022066    11
3733019363    11
3605634351    10
              ..
3736603529     1
3736386954     1
3736348678     1
3735669396     1
3767944844     1
Name: count, Length: 12010, dtype: int64

In [1824]:
df['Job ID'].value_counts().unique()

array([13, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1], dtype=int64)

In [1825]:
duplicates_count=df.duplicated(subset=['Job ID']).sum()
duplicates_count

3865

In [1826]:
duplicates = df['Job ID'].duplicated()
df[duplicates]

Job Title  \
603                             Odoo Developer   
781    Assistant Manager Sales/Sales Executive   
834                  Back-End Network Engineer   
840                Release Management Engineer   
842                                        NaN   
...                                        ...   
15856     Assistant Manager Talent Acquisition   
15857                        Software Engineer   
15858                    Web Backend Developer   
15860                       Investment Advisor   
15871                 Release Support Engineer   

                                                Job Info  \
603    AKSiQ · Karachi Division, Sindh, Pakistan (On-...   
781    Eastern Housing Pakistan · Multan, Punjab, Pak...   
834    Turing · Pakistan (Remote) Reposted  \n2 weeks...   
840    AIIT Solutions · Rawalpindi, Punjab, Pakistan ...   
842                                                  NaN   
...                                                  ...   
15856  Zameen.com · Lahore, Punjab, Pakistan  \n2 wee...   
15857    Arrivy · Lahore, Punjab, Pakistan  \n1 week ago   
15858  Turing · Pakistan  \n2 weeks ago\n\n · 11 appl...   
15860  Opal smart · Islamabad, Islāmābād, Pakistan  \...   
15871           Turing · Pakistan Reposted  \n1 week ago   

                                             Job Type  \
603                                         Full-time   
781                                         Full-time   
834                           Full-time · Entry level   
840                                         Full-time   
842                                               NaN   
...                                               ...   
15856  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15857                         On-site\n\n \nFull-time   
15858        Remote\n\n \nFull-time\n\n \nEntry level   
15860                         On-site\n\n \nFull-time   
15871        Remote\n\n \nFull-time\n\n \nEntry level   

                                            Job Link  \
603    https://www.linkedin.com/jobs/view/3727054738   
781    https://www.linkedin.com/jobs/view/3727057147   
834    https://www.linkedin.com/jobs/view/3726337380   
840    https://www.linkedin.com/jobs/view/3727054443   
842    https://www.linkedin.com/jobs/view/3727051751   
...                                              ...   
15856  https://www.linkedin.com/jobs/view/3765449305   
15857  https://www.linkedin.com/jobs/view/3765793064   
15858  https://www.linkedin.com/jobs/view/3765780780   
15860  https://www.linkedin.com/jobs/view/3766397564   
15871  https://www.linkedin.com/jobs/view/3766809535   

                                               employees      Job ID  
603                                      11-50 employees  3727054738  
781                                    201-500 employees  3727057147  
834    1,001-5,000 employees · IT Services and IT Con...  3726337380  
840                                      11-50 employees  3727054443  
842                                                  NaN  3727051751  
...                                                  ...         ...  
15856               5,001-10,000 employees · Real Estate  3765449305  
15857                                   51-200 employees  3765793064  
15858  1,001-5,000 employees · IT Services and IT Con...  3765780780  
15860      Areesha Rashid Qureshi is hiring for this job  3766397564  
15871  1,001-5,000 employees · IT Services and IT Con...  3766809535  

[3865 rows x 6 columns]

### drop duplicates

In [1827]:
df = df.drop_duplicates(subset=['Job ID'])

In [1828]:
df.shape

(12010, 6)

### checking null values

In [9]:
df.isnull().sum()

Job Title    288
Job Info     288
Job Type     302
Job Link       0
employees    303
Job ID         0
dtype: int64

In [10]:
null_rows = df[df.isnull().any(axis=1)]
null_rows

Job Title Job Info Job Type  \
110         NaN      NaN      NaN   
125         NaN      NaN      NaN   
417         NaN      NaN      NaN   
420         NaN      NaN      NaN   
456         NaN      NaN      NaN   
...         ...      ...      ...   
15787       NaN      NaN      NaN   
15789       NaN      NaN      NaN   
15793       NaN      NaN      NaN   
15797       NaN      NaN      NaN   
15842       NaN      NaN      NaN   

                                                Job Link employees      Job ID  
110        https://www.linkedin.com/jobs/view/3716079439       NaN  3716079439  
125        https://www.linkedin.com/jobs/view/3728144603       NaN  3728144603  
417    https://www.linkedin.com/jobs/view/3728021212/...       NaN  3728021212  
420    https://www.linkedin.com/jobs/view/3727051751/...       NaN  3727051751  
456    https://www.linkedin.com/jobs/view/3726689376/...       NaN  3726689376  
...                                                  ...       ...         ...  
15787      https://www.linkedin.com/jobs/view/3772859487       NaN  3772859487  
15789      https://www.linkedin.com/jobs/view/3766809507       NaN  3766809507  
15793      https://www.linkedin.com/jobs/view/3766806694       NaN  3766806694  
15797      https://www.linkedin.com/jobs/view/3770354315       NaN  3770354315  
15842      https://www.linkedin.com/jobs/view/3767390372       NaN  3767390372  

[303 rows x 6 columns]

In [11]:
len(null_rows)

303

### dealing Null values

In [12]:
df= df.dropna()
df

Job Title  \
0                              LinkedIn Data Extractor   
1                                 It Management Intern   
2                               SQL Database Developer   
3                      Business Intelligence Developer   
4                           Web Developer - Internship   
...                                                ...   
15870  Manager – Growth Chat & SMS (Digital Marketing)   
15871                         Release Support Engineer   
15872                  Information Security Specialist   
15873        FUNCTIONAL CONSULTANT ORACLE FUSION - HCM   
15874                              Community Associate   

                                                Job Info  \
0            EcomFleet · Pakistan (Remote)  \n1 week ago   
1      Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2      Murkez Technologies · Lahore, Punjab, Pakistan...   
3      DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4      ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                  ...   
15870  Bank Alfalah Limited · Karāchi, Sindh, Pakista...   
15871           Turing · Pakistan Reposted  \n1 week ago   
15872  DPL · Islamabad, Islāmābād, Pakistan  \n1 week...   
15873  Level 3 BOS · Karachi Division, Sindh, Pakista...   
15874  The Hive · Karachi Division, Sindh, Pakistan  ...   

                                             Job Type  \
0                                           Full-time   
1                                           Full-time   
2                                           Full-time   
3                                           Full-time   
4                                           Full-time   
...                                               ...   
15870  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15871        Remote\n\n \nFull-time\n\n \nEntry level   
15872       On-site\n\n \nFull-time\n\n \nEntry level   
15873  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15874                         On-site\n\n \nFull-time   

                                            Job Link  \
0      https://www.linkedin.com/jobs/view/3727054368   
1      https://www.linkedin.com/jobs/view/3723395858   
2      https://www.linkedin.com/jobs/view/3727034665   
3      https://www.linkedin.com/jobs/view/3727029273   
4      https://www.linkedin.com/jobs/view/3727034696   
...                                              ...   
15870  https://www.linkedin.com/jobs/view/3767355025   
15871  https://www.linkedin.com/jobs/view/3766809535   
15872  https://www.linkedin.com/jobs/view/3772530977   
15873  https://www.linkedin.com/jobs/view/3771943212   
15874  https://www.linkedin.com/jobs/view/3767944844   

                                               employees      Job ID  
0                                        11-50 employees  3727054368  
1                                        11-50 employees  3723395858  
2                                       51-200 employees  3727034665  
3                                        11-50 employees  3727029273  
4                    Mobeen Naqvi is hiring for this job  3727034696  
...                                                  ...         ...  
15870                   5,001-10,000 employees · Banking  3767355025  
15871  1,001-5,000 employees · IT Services and IT Con...  3766809535  
15872   51-200 employees · IT Services and IT Consulting  3772530977  
15873  201-500 employees · Outsourcing and Offshoring...  3771943212  
15874                                    11-50 employees  3767944844  

[15572 rows x 6 columns]

### drop duplicates

In [13]:
# df = df.drop_duplicates(subset=['Job ID'])

In [14]:
# df.shape

(11744, 6)

## to be noted that
***drop null and then drop duplicates:11744 rows***

***drop duplicates and then drop null :11737 rows***

### exploring values in features

In [1834]:
df['Job Title'].unique()	

array(['LinkedIn Data Extractor', 'It Management Intern',
       'SQL Database Developer', ...,
       'Manager – Growth Chat & SMS (Digital Marketing)',
       'Information Security Specialist',
       'FUNCTIONAL CONSULTANT ORACLE FUSION - HCM'], dtype=object)

In [1835]:
len(df['Job Title'].unique())

6442

In [1836]:
df['Job Title'].value_counts()[:30]

Job Title
Sales Executive                          142
Business Development Executive           135
Sales Specialist                          76
Business Development Manager              71
Business Development Specialist           63
Graphic Designer                          58
Web Back-End Developer                    57
Software Engineer                         56
Customer Service Representative           55
Salesperson                               50
Full Stack Engineer                       47
Senior Software Engineer                  45
Sales And Marketing Specialist            40
Wordpress Developer                       38
Call Center Representative                37
PHP Developer                             36
Customer Sales Representative             35
Sales Development Representative          35
Search Engine Optimization Specialist     33
Dotnet Developer                          33
Digital Marketing Specialist              32
Business Development Officer              32


In [1837]:
df['Job Info'].unique()

array(['EcomFleet · Pakistan (Remote)  \n1 week ago',
       'Digicon Valley · Lahore, Punjab, Pakistan (On-site)  \n2 weeks ago',
       'Murkez Technologies · Lahore, Punjab, Pakistan (On-site)  \n1 week ago',
       ...,
       'DPL · Islamabad, Islāmābād, Pakistan  \n1 week ago\n\n · 83 applicants',
       'Level 3 BOS · Karachi Division, Sindh, Pakistan  \n1 week ago\n\n · 19 applicants',
       'The Hive · Karachi Division, Sindh, Pakistan  \n1 week ago'],
      dtype=object)

In [1838]:
len(df['Job Info'].unique())

10209

In [1839]:
df['Job Type'].unique()

array(['Full-time', 'Full-time · Mid-Senior level',
       'Internship · Internship', 'Full-time · Associate', 'Internship',
       'Full-time · Entry level', 'Part-time', 'Full-time · Internship',
       'Contract · Mid-Senior level', 'Contract', 'Contract · Associate',
       '$150/month - $300/month · Full-time', 'Full-time · Director',
       'Part-time · Associate', 'Mr Ahsan is hiring for this job',
       'Part-time · Executive', 'Contract · Executive',
       'Part-time · Mid-Senior level', 'Contract · Entry level',
       'Temporary · Mid-Senior level', 'Executive',
       'Full-time · Executive', 'Sean Sato is hiring for this job',
       'Part-time · Internship', 'Internship · Mid-Senior level',
       '$12,000/yr - $15,000/yr · Full-time · Mid-Senior level',
       '$200,000/yr · Full-time', 'Part-time · Entry level',
       'dr.ry. L. is hiring for this job',
       '10,001+ employees · Tobacco Manufacturing', 'Volunteer',
       'On-site\n\n \nFull-time\n\n \nMid-Senior l

In [1840]:
len(df['Job Type'].unique())

129

In [1841]:
df['employees']

0                                          11-50 employees
1                                          11-50 employees
2                                         51-200 employees
3                                          11-50 employees
4                      Mobeen Naqvi is hiring for this job
                               ...                        
15869                                       1-10 employees
15870                     5,001-10,000 employees · Banking
15872     51-200 employees · IT Services and IT Consulting
15873    201-500 employees · Outsourcing and Offshoring...
15874                                      11-50 employees
Name: employees, Length: 11737, dtype: object

In [1842]:
df['employees'].unique()[:10]

array(['11-50 employees', '51-200 employees',
       'Mobeen Naqvi is hiring for this job',
       '1,001-5,000 employees · Software Development',
       '501-1,000 employees · IT Services and IT Consulting',
       '501-1,000 employees · Telecommunications',
       '1,001-5,000 employees · Banking', '201-500 employees',
       '10,001+ employees', '51-200 employees · Human Resources Services'],
      dtype=object)

In [1843]:
len(df['employees'].unique())

975

### extracting company feature

In [1844]:
df.loc[:,'company']=df['Job Info'].apply(lambda x:x.split('·')[0])

In [1845]:
df['company'].unique()

array(['EcomFleet ', 'Digicon Valley ', 'Murkez Technologies ', ...,
       'EComia Global ', 'BUZZMAX.CA ', 'Codiea.io '], dtype=object)

In [1846]:
len(df['company'].unique())

4435

In [1847]:
df['company'].value_counts()

company
Crossover                                   848
Zameen.com                                  252
Turing                                      223
People (Professional Employers Pvt Ltd)     143
Zones IT Solutions                          116
                                           ... 
Smart IT Solutions                            1
Al Falah Chemicals                            1
Office Automation Group                       1
Jana.Pk - Travel & Tour Services              1
Codiea.io                                     1
Name: count, Length: 4435, dtype: int64

In [1848]:
df['company'].value_counts()[:30]

company
Crossover                                         848
Zameen.com                                        252
Turing                                            223
People (Professional Employers Pvt Ltd)           143
Zones IT Solutions                                116
Afiniti                                           104
Jazz                                              101
Times Consultant (Pvt.) Ltd.                       88
Daraz                                              79
Telenor                                            70
HR Ways - Hiring Tech Talent                       67
The Indus Hospital                                 66
CureMD                                             66
Motive                                             65
PTCL.Official                                      59
Kamayi                                             57
BAT                                                56
K-Electric                                         49
Dubizzle Labs       

### extracting location

In [1849]:
df.loc[:,'location']=df['Job Info'].apply(lambda x:x.split('·')[1].split()[0].strip(','))

In [1850]:
df['location']

0          Pakistan
1            Lahore
2            Lahore
3         Islamabad
4            Lahore
            ...    
15869    Bahawalpur
15870       Karāchi
15872     Islamabad
15873       Karachi
15874       Karachi
Name: location, Length: 11737, dtype: object

In [1851]:
df['location'].unique()

array(['Pakistan', 'Lahore', 'Islamabad', 'Karāchi', 'Peshawar',
       'Islāmābād', 'Rawalpindi', 'Multan', 'Karachi', 'Hyderabad',
       'Mirpur', 'Kharian', 'Sindh', 'Faisalabad', 'Jhelum', 'Gujrāt',
       'Korangi', 'Bahawalpur', 'Gulshan', 'Gujranwala', 'Sukkur',
       'Punjab', 'Shekhūpura', 'Sargodha', 'Wazirabad', 'Mānsehra',
       'Sialkot', 'Chakwāl', 'Muzaffargarh', 'Nawabshah', 'Mardān',
       'Gujrat', 'Dera', 'Sāhīwāl', 'Jhang', 'Sahiwal', 'Quetta',
       'Vienna', 'Rahimyar', 'Milan', 'Lodhran', 'South', 'Zhob',
       'Jamshoro', 'Khyber', 'New', 'Rakhni', 'Taxila', 'APAC', 'Azad',
       'Toba', 'Thatta', 'Jarānwāla', 'Mīrpur', 'Larkana', 'Pasrur',
       'Sarai', 'Panjgur', 'Baloch', 'Attock', 'Rahim', 'Ratodero', 'APJ',
       'Chakwal', 'İzmir', 'Miān', 'Battagram', 'Harīpur', 'Naushahro',
       'Khuzdar', 'Kasur', 'Abbottabad', 'Malir', 'Vehari', 'Kahuta',
       'Badin', 'Thal', 'Landhi', 'Shikarpur', 'Bhakkar', 'Bahrain',
       'Metro', 'Kohāt', 'Okāra', 

In [1852]:
len(df['location'].unique())

149

In [1853]:
df['location'].value_counts()

location
Lahore          3842
Karāchi         2274
Islamabad       1725
Pakistan         900
Karachi          790
                ... 
Sibi               1
Bin                1
Shergarh           1
Bahawalnagar       1
Sheikhupura        1
Name: count, Length: 149, dtype: int64

In [1854]:
df['location'].value_counts()[:30]

location
Lahore          3842
Karāchi         2274
Islamabad       1725
Pakistan         900
Karachi          790
Rawalpindi       551
Islāmābād        309
Faisalabad       202
Multan           124
Peshawar         105
Punjab            84
Hyderabad         69
Sialkot           60
APAC              51
Gujranwala        49
APJ               38
Quetta            37
New               34
Sindh             32
South             30
Bahawalpur        29
Muzaffargarh      28
Jhelum            22
Gujrāt            21
Sargodha          15
Sukkur            14
Rahim             13
Jamshoro          12
Umarkot           11
Sahiwal           11
Name: count, dtype: int64

In [1855]:
df.loc[:,'location']=df['Job Info'].apply(lambda x:x.split('·')[1].split('(')[0])

In [1856]:
df['location']

0                                                Pakistan 
1                                Lahore, Punjab, Pakistan 
2                                Lahore, Punjab, Pakistan 
3                          Islamabad, Islāmābād, Pakistan 
4                                Lahore, Punjab, Pakistan 
                               ...                        
15869           Bahawalpur, Punjab, Pakistan  \n1 week ago
15870               Karāchi, Sindh, Pakistan  \n1 week ago
15872     Islamabad, Islāmābād, Pakistan  \n1 week ago\...
15873     Karachi Division, Sindh, Pakistan  \n1 week a...
15874      Karachi Division, Sindh, Pakistan  \n1 week ago
Name: location, Length: 11737, dtype: object

In [1857]:
df['location'].unique()

array([' Pakistan ', ' Lahore, Punjab, Pakistan ',
       ' Islamabad, Islāmābād, Pakistan ', ...,
       ' Sakrand, Sindh, Pakistan  \n1 week ago\n\n ',
       ' Sheikhupura District, Punjab, Pakistan  \n1 week ago\n\n ',
       ' Islamabad, Punjab, Pakistan  \n2 weeks ago\n\n '], dtype=object)

In [1858]:
len(df['location'].unique())

1163

In [1859]:
df['location'].value_counts()

location
 Lahore, Punjab, Pakistan                                  1052
 Karāchi, Sindh, Pakistan                                   624
 Islamabad, Islāmābād, Pakistan                             518
 Lahore, Punjab, Pakistan  \n2 weeks ago                    326
 Karachi Division, Sindh, Pakistan                          242
                                                           ... 
 Gulshan Town, Sindh, Pakistan  \n7 hours ago\n\n             1
 Korangi District, Sindh, Pakistan  \n10 hours ago\n\n        1
 Jhelum, Punjab, Pakistan  \n23 hours ago\n\n                 1
 Sāhīwāl, Punjab, Pakistan  \n9 hours ago\n\n                 1
 Islamabad, Punjab, Pakistan  \n2 weeks ago\n\n               1
Name: count, Length: 1163, dtype: int64

In [1860]:
df['location'].value_counts()[:30]

location
 Lahore, Punjab, Pakistan                              1052
 Karāchi, Sindh, Pakistan                               624
 Islamabad, Islāmābād, Pakistan                         518
 Lahore, Punjab, Pakistan  \n2 weeks ago                326
 Karachi Division, Sindh, Pakistan                      242
 Pakistan                                               236
 Lahore, Punjab, Pakistan  \n2 weeks ago\n\n            199
 Lahore, Punjab, Pakistan  \n1 week ago                 194
 Lahore, Punjab, Pakistan  \n1 week ago\n\n             177
 Karāchi, Sindh, Pakistan  \n1 week ago\n\n             145
 Lahore, Punjab, Pakistan  \n1 month ago                138
 Karāchi, Sindh, Pakistan  \n2 weeks ago                138
 Rawalpindi, Punjab, Pakistan                           127
 Karāchi, Sindh, Pakistan  \n2 weeks ago\n\n            127
 Islamabad, Islāmābād, Pakistan  \n2 weeks ago\n\n      126
 Pakistan  \n2 weeks ago\n\n                            124
 Islamabad, Islāmābād, Pakistan

### extracting remote/onsite

In [1861]:
remote_onsite_list=[]
for info in df['Job Info']:
    try:
        remote_onsite_list.append(info.split('·')[1].split('(')[1].split(')')[0])
    except IndexError:
        remote_onsite_list.append('Not specified')
    

In [1862]:
len(remote_onsite_list)

11737

In [1863]:
df.loc[:,'Remote/Onsite']=remote_onsite_list

In [1864]:
df['Remote/Onsite']

0               Remote
1              On-site
2              On-site
3               Hybrid
4              On-site
             ...      
15869    Not specified
15870    Not specified
15872    Not specified
15873    Not specified
15874    Not specified
Name: Remote/Onsite, Length: 11737, dtype: object

In [1865]:
df['Remote/Onsite'].unique()

array(['Remote', 'On-site', 'Hybrid', 'Not specified'], dtype=object)

In [1866]:
df['Remote/Onsite'].value_counts()

Remote/Onsite
Not specified    8396
On-site          2616
Remote            525
Hybrid            200
Name: count, dtype: int64

### extarctig number of applicants feature

In [1867]:
n_applicant_list=[]
for info in df['Job Info']:
    try:
        n_applicant_list.append(info.split('·')[2].strip(' applicants'))
    except IndexError:
        n_applicant_list.append(np.nan)

In [1868]:
n_applicant_list[:10]

[nan, nan, nan, '51', '51', nan, nan, nan, '312', '2,048']

In [1869]:
df.loc[:,'n_applicant']=n_applicant_list

In [1870]:
df['n_applicant'].unique()

array([nan, '51', '312', '2,048', '30', '3,159', '41', '3,008', '120',
       '685', '50', '28', '24', '734', '9', '112', '42', '10', '89', '48',
       '19', '732', '338', '18', '154', '472', '110', '32', '16', '448',
       '239', '933', '255', '15', '177', '33', '323', '207', '22', '21',
       '49', '591', '45', '29', '231', '337', '249', '123', '40', '192',
       '157', '1,386', '36', '1,186', '17', '76', '61', '38', '20', '12',
       '115', '73', '46', '315', '90', '67', '27', '23', '13', '172',
       '778', '96', '58', '216', '43', '204', '247', '380', '39', '1,768',
       '11', '310', '34', '108', '162', '853', '986', '215', '14', '94',
       '92', '168', '911', '53', '420', '31', '72', '44', '59', '132',
       '289', '35', '52', '1,379', '158', '260', '54', '250', '332',
       '649', '1,385', '153', '37', '83', '370', '287', '57', '339', '47',
       '208', '121', '56', '98', '68', '254', '259', '138', '305', '186',
       '197', '84', '795', '63', '79', '757', '166', '

In [1871]:
len(df['n_applicant'].unique())

756

In [1872]:
df['n_applicant'].value_counts()[:30]

n_applicant
51          287
Over 200    271
11          194
13          140
10          135
12          132
0           129
15          122
14          121
18          108
17          107
19          103
1           101
16           98
9            94
3            93
21           90
7            89
8            89
2            85
4            83
6            82
20           79
5            78
23           68
24           67
25           67
32           66
26           61
22           59
Name: count, dtype: int64

### Extracting job posting time feature

In [1873]:
df.loc[:,'job posting time']=df['Job Info'].apply(lambda x:x.split('\n')[1])
df['job posting time']

0         1 week ago
1        2 weeks ago
2         1 week ago
3         1 week ago
4         1 week ago
            ...     
15869     1 week ago
15870     1 week ago
15872     1 week ago
15873     1 week ago
15874     1 week ago
Name: job posting time, Length: 11737, dtype: object

In [1874]:
df['job posting time'].unique()

array(['1 week ago', '2 weeks ago', '1 day ago', '6 days ago',
       '5 days ago', '11 hours ago', '3 days ago', '3 hours ago',
       '4 days ago', '13 minutes ago', '14 minutes ago', '5 hours ago',
       '10 hours ago', '9 hours ago', '2 days ago', '13 hours ago',
       '18 hours ago', '7 hours ago', '4 hours ago', '3 weeks ago',
       '2 hours ago', '12 hours ago', '6 hours ago', '14 hours ago',
       '1 hour ago', '22 hours ago', '8 hours ago', '23 hours ago',
       '15 hours ago', '20 hours ago', '19 hours ago', '16 hours ago',
       '17 hours ago', '21 hours ago', '17 minutes ago', '48 minutes ago',
       '23 minutes ago', '49 minutes ago', '37 minutes ago',
       '42 minutes ago', '36 minutes ago', '44 minutes ago',
       '55 minutes ago', '32 minutes ago', '58 minutes ago',
       '3 minutes ago', '1 month ago', '4 weeks ago',
       'https://anitechnologies.net/ · Islamabad, Islāmābād, Pakistan Reposted  '],
      dtype=object)

In [1875]:
df['job posting time'].value_counts()[:10]

job posting time
1 week ago     3578
2 weeks ago    2243
5 days ago      854
6 days ago      743
3 days ago      664
1 day ago       528
1 month ago     495
4 days ago      491
2 days ago      474
3 weeks ago     319
Name: count, dtype: int64

### Feature Eng on Job Type feature

In [1876]:
df['Job Type'].unique()

array(['Full-time', 'Full-time · Mid-Senior level',
       'Internship · Internship', 'Full-time · Associate', 'Internship',
       'Full-time · Entry level', 'Part-time', 'Full-time · Internship',
       'Contract · Mid-Senior level', 'Contract', 'Contract · Associate',
       '$150/month - $300/month · Full-time', 'Full-time · Director',
       'Part-time · Associate', 'Mr Ahsan is hiring for this job',
       'Part-time · Executive', 'Contract · Executive',
       'Part-time · Mid-Senior level', 'Contract · Entry level',
       'Temporary · Mid-Senior level', 'Executive',
       'Full-time · Executive', 'Sean Sato is hiring for this job',
       'Part-time · Internship', 'Internship · Mid-Senior level',
       '$12,000/yr - $15,000/yr · Full-time · Mid-Senior level',
       '$200,000/yr · Full-time', 'Part-time · Entry level',
       'dr.ry. L. is hiring for this job',
       '10,001+ employees · Tobacco Manufacturing', 'Volunteer',
       'On-site\n\n \nFull-time\n\n \nMid-Senior l

In [1877]:
len(df['Job Type'].unique())

129

## understanding options according to linkedin
****

### These are job Job type options on linkedin:
Full-time,
Part-time,
Contract,
Temporary,
Volunteer,
Internship,
Other
### These are Experience level options on linkedin:
Internship,
Entry level,
Associate,
Mid-Senior level,
Director,
Executive
### These are On-site/remote options on linkedin:
On-site,
Remote,
Hybrid

***

### counting number of elements in job type column

In [1878]:
Job_Type_nElements=[]
for cell in df['Job Type']:
    if '·' in cell:
        nElements=len(cell.split('·'))
        Job_Type_nElements.append(nElements)
    elif '\n\n \n' in cell:
        nElements=len(cell.split('\n\n \n'))
        Job_Type_nElements.append(nElements)
    else:
        nElements=len(cell.split())
        Job_Type_nElements.append(nElements)        

In [1879]:
set(Job_Type_nElements) 

{1, 2, 3, 4, 7}

In [1880]:
df.loc[:,'Job Type nElements']=Job_Type_nElements

In [1881]:
df['Job Type nElements'].value_counts()

Job Type nElements
2    5693
3    3633
1    2395
4      13
7       3
Name: count, dtype: int64

In [1882]:
df[df['Job Type nElements']==4]
df[df['Job Type nElements']==4]['Job Type']

4980     PKR600,000/yr - PKR660,000/yr\n\n \nOn-site\n\...
8942     PKR1,200,000/yr - PKR1,800,000/yr\n\n \nOn-sit...
8961     PKR1,200,000/yr - PKR2,400,000/yr\n\n \nOn-sit...
9066     PKR1,800,000/yr - PKR3,600,000/yr\n\n \nOn-sit...
9119     MYR57,000/yr - MYR108,000/yr\n\n \nRemote\n\n ...
10276    PHP360,000/yr - PHP720,000/yr\n\n \nRemote\n\n...
11404    PKR960,000/yr - PKR1,200,000/yr\n\n \nOn-site\...
11506    $20,000/yr - $25,000/yr\n\n \nRemote\n\n \nCon...
12575    PKR960,000/yr - PKR1,200,000/yr\n\n \nOn-site\...
14039    $36,000/yr - $60,000/yr\n\n \nRemote\n\n \nFul...
14391    ₹180,000/yr - ₹360,000/yr\n\n \nRemote\n\n \nF...
14491    PHP360,000/yr - PHP720,000/yr\n\n \nRemote\n\n...
15090    $12,000/yr - $18,000/yr\n\n \nRemote\n\n \nFul...
Name: Job Type, dtype: object

In [1883]:
df[df['Job Type nElements']==4]['Job Type'].iloc[0]
df[df['Job Type nElements']==4]['Job Type'].iloc[0].split()
len(df[df['Job Type nElements']==4]['Job Type'].iloc[0].split())
# df[df['Job Type nElements']==4]['Job Type'].iloc[0].split('\n\n \n')

7

### extarcting Job type

In [1884]:
inconsistant_Job_Type_Elements_1=['Executive']
inconsistant_Job_Type_Elements_1

['Executive']

In [1885]:
inconsistant_Job_Type_Elements_2=['$150/month - $300/month · Full-time','$200,000/yr · Full-time','10,001+ employees · Tobacco Manufacturing','Full-time\n\n \nMid-Senior level']
inconsistant_Job_Type_Elements_2

['$150/month - $300/month · Full-time',
 '$200,000/yr · Full-time',
 '10,001+ employees · Tobacco Manufacturing',
 'Full-time\n\n \nMid-Senior level']

In [1886]:
inconsistant_Job_Type_Elements_3=['$12,000/yr - $15,000/yr · Full-time · Mid-Senior level']
inconsistant_Job_Type_Elements_3

['$12,000/yr - $15,000/yr · Full-time · Mid-Senior level']

In [1887]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split())==1:
        if elmnt not in inconsistant_Job_Type_Elements_1:
            Job_Type_list.append(elmnt)
        else:
            Job_Type_list.append('Not specified')
len(Job_Type_list)

2395

In [1888]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split('·'))==2 or len(elmnt.split('\n\n \n'))==2:
        if elmnt not in inconsistant_Job_Type_Elements_2:
            if '·' in elmnt:
                Job_Type_list.append(elmnt.split('·')[0])
            elif '\n\n \n' in elmnt:
                Job_Type_list.append(elmnt.split('\n\n \n')[1])
            else:
                Job_Type_list.append('Not specified')
        else:
            if 'Full-time' in elmnt:
                Job_Type_list.append('Full-time')
            else:
                Job_Type_list.append('Not specified')

len(Job_Type_list) # problem it should be:5693

5692

In [1889]:
inconsistant_Job_Type_Elements_2

['$150/month - $300/month · Full-time',
 '$200,000/yr · Full-time',
 '10,001+ employees · Tobacco Manufacturing',
 'Full-time\n\n \nMid-Senior level']

In [1890]:
df[df['Job Type nElements']==2]['Job Type'].unique()

array(['Full-time · Mid-Senior level', 'Internship · Internship',
       'Full-time · Associate', 'Full-time · Entry level',
       'Full-time · Internship', 'Contract · Mid-Senior level',
       'Contract · Associate', '$150/month - $300/month · Full-time',
       'Full-time · Director', 'Part-time · Associate',
       'Part-time · Executive', 'Contract · Executive',
       'Part-time · Mid-Senior level', 'Contract · Entry level',
       'Temporary · Mid-Senior level', 'Full-time · Executive',
       'Part-time · Internship', 'Internship · Mid-Senior level',
       '$200,000/yr · Full-time', 'Part-time · Entry level',
       '10,001+ employees · Tobacco Manufacturing',
       'On-site\n\n \nFull-time', 'Remote\n\n \nFull-time',
       'Full-time\n\n \nMid-Senior level', 'On-site\n\n \nPart-time',
       'Hybrid\n\n \nFull-time', 'Remote\n\n \nInternship',
       'On-site\n\n \nInternship', 'Remote\n\n \nContract',
       'Remote\n\n \nPart-time', 'Full-time\n\n \nEntry level',
       

In [1891]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split('·'))==3 or len(elmnt.split('\n\n \n'))==3:
        if elmnt not in inconsistant_Job_Type_Elements_3:
            Job_Type_list.append(elmnt.split('\n\n \n')[1])
        else:
            if 'Full-time' in elmnt:
                Job_Type_list.append('Full-time')
            else:
                Job_Type_list.append('Not specified')
len(Job_Type_list)

3633

In [1892]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split())==7:
        Job_Type_list.append('Not specified')
len(Job_Type_list) # problem it should be 3

18

In [1893]:
13+3 # it hase two extra values

16

In [1894]:
df['Job Type nElements'].value_counts()

Job Type nElements
2    5693
3    3633
1    2395
4      13
7       3
Name: count, dtype: int64

In [1895]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split())==1:
        if elmnt not in inconsistant_Job_Type_Elements_1:
            Job_Type_list.append(elmnt)
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==2 or len(elmnt.split('\n\n \n'))==2:
        if elmnt not in inconsistant_Job_Type_Elements_2:
            if '·' in elmnt:
                Job_Type_list.append(elmnt.split('·')[0])
            elif '\n\n \n' in elmnt:
                Job_Type_list.append(elmnt.split('\n\n \n')[1])
            else:
                Job_Type_list.append('Not specified')
        else:
            if 'Full-time' in elmnt:
                Job_Type_list.append('Full-time')
            else:
                Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==3 or len(elmnt.split('\n\n \n'))==3:
        if elmnt not in inconsistant_Job_Type_Elements_3:
            Job_Type_list.append(elmnt.split('\n\n \n')[1])
        else:
            if 'Full-time' in elmnt:
                Job_Type_list.append('Full-time')
            else:
                Job_Type_list.append('Not specified')
    if len(elmnt.split())==7:
        Job_Type_list.append('Not specified')

In [1896]:
len(Job_Type_list)

11738

In [1897]:
Job_type_options=['Internship','Full-time','Part-time','Contract','Temporary','Volunteer','Other']
Job_type_options
# len(Job_type_options)
# 'Internship' placed at first due to avoid collision in experience level

['Internship',
 'Full-time',
 'Part-time',
 'Contract',
 'Temporary',
 'Volunteer',
 'Other']

### new way to extract Job-Type

In [1898]:
filtered_df = df[df['Job Type'].apply(lambda job_type: any(keyword in job_type for keyword in Job_type_options))]
filtered_df
len(filtered_df)

11712

In [1899]:
len(df)-len(filtered_df)

25

In [1900]:
x=lambda job_type: any(keyword in job_type for keyword in Job_type_options)
x('Full-time · Mid-Senior level')                       

True

In [1901]:
#we can write the above lambda function by this way
# def check_keywords(job_type, keyword_options):
#     for keyword in keyword_options:
#         if keyword in job_type:
#             return True
#     return False

# # Example usage:
# Job_type_options = ["keyword1", "keyword2", "keyword3"]
# job_type = "Some job title with keyword2"

# result = check_keywords(job_type, Job_type_options)
# print(result)
job_info = "Full-time · Mid-Senior level"
job_type_list=[]
job_type=None
flag=0
for keyword in Job_type_options:
    if keyword in job_info:
        job_type=keyword
        flag=1
        # break
    else:
        pass
        # flag=0
if flag==1:
    job_type_list.append(job_type)
else:
    job_type_list.append('Not specified')
            

job_type_list

['Full-time']

In [1902]:
Job_Type_list=[]
for i in range(len(df['Job Type'].iloc[:])):
    job_type=None
    flag=0
    for keyword in Job_type_options:
        if keyword in df['Job Type'].iloc[i]:
            job_type=keyword
            flag=1
            # break
        else:
            pass
            # flag=0
    if flag==1:
        Job_Type_list.append(job_type)
    else:
        Job_Type_list.append('Not specified')

Job_Type_list
len(Job_Type_list)

11737

In [1903]:
set(Job_Type_list)

{'Contract',
 'Full-time',
 'Internship',
 'Not specified',
 'Part-time',
 'Temporary',
 'Volunteer'}

#### for verification

In [1904]:
Job_Type_list.count('Not specified')

25

In [1905]:
len(df)-Job_Type_list.count('Not specified')==len(df)-Job_Type_list.count('Not specified')

True

In [1906]:
11737-25

11712

In [1907]:
df[df['Job Type']=='10,001+ employees · Food and Beverage Manufacturing'].iloc[0] #'Part-time · Mid-Senior level'

Job Title                                             My Pay Lead MENAP
Job Info              Mondelēz International · Karāchi, Sindh, Pakis...
Job Type              10,001+ employees · Food and Beverage Manufact...
Job Link                  https://www.linkedin.com/jobs/view/3746235031
employees                                     1 company alum works here
Job ID                                                       3746235031
company                                         Mondelēz International 
location                 Karāchi, Sindh, Pakistan  \n3 minutes ago\n\n 
Remote/Onsite                                             Not specified
n_applicant                                                           8
job posting time                                          3 minutes ago
Job Type nElements                                                    2
Name: 9389, dtype: object

***working good***

In [1908]:
df.loc[:,'Job-Type']=Job_Type_list

In [1909]:
df['Job-Type'].unique()


array(['Full-time', 'Internship', 'Part-time', 'Contract',
       'Not specified', 'Temporary', 'Volunteer'], dtype=object)

In [1910]:
len(df['Job-Type'].unique())

7

In [1911]:
df['Job-Type'].value_counts()

Job-Type
Full-time        11073
Contract           277
Part-time          211
Internship         128
Not specified       25
Temporary           18
Volunteer            5
Name: count, dtype: int64

### extarcting Experience level from df['Job Type']

In [1912]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split())==1:
        if elmnt in inconsistant_Job_Type_Elements_1:
            if elmnt=='Executive':
                Job_Type_list.append(elmnt)
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==2 or len(elmnt.split('\n\n \n'))==2:
        if elmnt not in inconsistant_Job_Type_Elements_2:
            if '·' in elmnt:
                Job_Type_list.append(elmnt.split('·')[1])
            # elif '\n\n \n' in elmnt:
            #     Job_Type_list.append(elmnt.split('\n\n \n')[2])
            else:
                Job_Type_list.append('Not specified')
        else:
            if 'Mid-Senior level' in elmnt:
                Job_Type_list.append('Mid-Senior level')
            else:
                Job_Type_list.append('Not specified')
                
    if len(elmnt.split('·'))==3 or len(elmnt.split('\n\n \n'))==3:
        if elmnt not in inconsistant_Job_Type_Elements_3:
            Job_Type_list.append(elmnt.split('\n\n \n')[2])
        else:
            # if 'Mid-Senior level' in inconsistant_Job: due to this problem not sepecified is in list 
            if 'Mid-Senior level' in elmnt:  
                Job_Type_list.append('Mid-Senior level')
            else:
                Job_Type_list.append('Not specified')
    if len(elmnt.split())==7:
        Job_Type_list.append('Not specified')

In [1913]:
len(Job_Type_list)

11738

In [1914]:
set(Job_Type_list)

{' Associate',
 ' Director',
 ' Entry level',
 ' Executive',
 ' Food and Beverage Manufacturing',
 ' IT Services and IT Consulting',
 ' Internship',
 ' Mid-Senior level',
 'Associate',
 'Contract',
 'Director',
 'Entry level',
 'Executive',
 'Full-time',
 'Internship',
 'Mid-Senior level',
 'Not specified'}

In [1915]:
Experience_level_options=['Internship','Entry level','Associate','Mid-Senior level','Director','Executive']
Experience_level_options
# len(Experience_level_options)

['Internship',
 'Entry level',
 'Associate',
 'Mid-Senior level',
 'Director',
 'Executive']

### new way to extract Experience Level

In [1916]:
df[df['Job Type nElements']==3]['Job Type']
df[df['Job Type nElements']==3]['Job Type'].unique()

array(['$12,000/yr - $15,000/yr · Full-time · Mid-Senior level',
       'On-site\n\n \nFull-time\n\n \nMid-Senior level',
       'Hybrid\n\n \nFull-time\n\n \nMid-Senior level',
       'Remote\n\n \nFull-time\n\n \nMid-Senior level',
       'On-site\n\n \nFull-time\n\n \nAssociate',
       'Hybrid\n\n \nFull-time\n\n \nEntry level',
       'Hybrid\n\n \nFull-time\n\n \nDirector',
       'On-site\n\n \nFull-time\n\n \nEntry level',
       'Hybrid\n\n \nFull-time\n\n \nAssociate',
       'Remote\n\n \nContract\n\n \nMid-Senior level',
       'On-site\n\n \nFull-time\n\n \nDirector',
       'On-site\n\n \nFull-time\n\n \nExecutive',
       'Remote\n\n \nFull-time\n\n \nEntry level',
       'On-site\n\n \nContract\n\n \nMid-Senior level',
       'Remote\n\n \nFull-time\n\n \nAssociate',
       'On-site\n\n \nContract\n\n \nEntry level',
       'On-site\n\n \nFull-time\n\n \nInternship',
       'Remote\n\n \nPart-time\n\n \nAssociate',
       'Remote\n\n \nContract\n\n \nEntry level',
     

In [1917]:
filtered_df = df[df['Job Type'].apply(lambda job_type: any(keyword in job_type for keyword in Experience_level_options))]
filtered_df
len(filtered_df)

5055

In [1918]:
len(df)-len(filtered_df)

6682

In [1919]:
Job_Type_list=[]
for i in range(len(df['Job Type'].iloc[:])):
    job_type=None
    flag=0
    for keyword in Experience_level_options:
        if keyword in df['Job Type'].iloc[i]:
            job_type=keyword
            flag=1
            # break
        else:
            pass
            # flag=0
    if flag==1:
        Job_Type_list.append(job_type)
    else:
        Job_Type_list.append('Not specified')

Job_Type_list
len(Job_Type_list)

11737

In [1920]:
set(Job_Type_list)

{'Associate',
 'Director',
 'Entry level',
 'Executive',
 'Internship',
 'Mid-Senior level',
 'Not specified'}

### for verification

In [1921]:
Job_Type_list.count('Not specified')

6682

In [1922]:
len(df)-Job_Type_list.count('Not specified') 
len(df)-Job_Type_list.count('Not specified')==len(df)-Job_Type_list.count('Not specified')

True

In [1923]:
df[df['Job Type']=='Hybrid\n\n \nInternship\n\n \nMid-Senior level'].iloc[0] #'$12,000/yr - $15,000/yr · Full-time · Mid-Senior level'

Job Title             Senior Consultant- Business Consulting- Intern...
Job Info              EY in Pakistan  · Karachi Division, Sindh, Pak...
Job Type                 Hybrid\n\n \nInternship\n\n \nMid-Senior level
Job Link                  https://www.linkedin.com/jobs/view/3767944882
employees                            1,001-5,000 employees · Accounting
Job ID                                                       3767944882
company                                                EY in Pakistan  
location               Karachi Division, Sindh, Pakistan  \n1 week a...
Remote/Onsite                                             Not specified
n_applicant                                                           6
job posting time                                             1 week ago
Job Type nElements                                                    3
Job-Type                                                     Internship
Name: 15868, dtype: object

***working good***

In [1924]:
df.loc[:,'Experience Level']=Job_Type_list

In [1925]:
df['Experience Level'].unique()

array(['Not specified', 'Mid-Senior level', 'Internship', 'Associate',
       'Entry level', 'Director', 'Executive'], dtype=object)

In [1926]:
len(df['Experience Level'].unique())

7

In [1927]:
df['Experience Level'].value_counts()

Experience Level
Not specified       6682
Mid-Senior level    2740
Entry level         1153
Associate            696
Director             212
Internship           152
Executive            102
Name: count, dtype: int64

### extracting On-site/remote from Job Type feature

In [1928]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split())==1:
        if elmnt not in inconsistant_Job_Type_Elements_1:
            Job_Type_list.append('Not specified')
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==2 or len(elmnt.split('\n\n \n'))==2:
        if elmnt not in inconsistant_Job_Type_Elements_2:
            if '·' in elmnt:
                Job_Type_list.append('Not specified')
            elif '\n\n \n' in elmnt:
                Job_Type_list.append(elmnt.split('\n\n \n')[0])
            else:
                Job_Type_list.append('Not specified')
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==3 or len(elmnt.split('\n\n \n'))==3:
        if elmnt not in inconsistant_Job_Type_Elements_3:
            Job_Type_list.append(elmnt.split('\n\n \n')[0])
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split())==7:
        Job_Type_list.append('Not specified')

### new way to extract Remote/Onsite-2

In [1929]:
df[df['Job Type nElements']==2]['Job Type']
df[df['Job Type nElements']==2]['Job Type'].unique()

array(['Full-time · Mid-Senior level', 'Internship · Internship',
       'Full-time · Associate', 'Full-time · Entry level',
       'Full-time · Internship', 'Contract · Mid-Senior level',
       'Contract · Associate', '$150/month - $300/month · Full-time',
       'Full-time · Director', 'Part-time · Associate',
       'Part-time · Executive', 'Contract · Executive',
       'Part-time · Mid-Senior level', 'Contract · Entry level',
       'Temporary · Mid-Senior level', 'Full-time · Executive',
       'Part-time · Internship', 'Internship · Mid-Senior level',
       '$200,000/yr · Full-time', 'Part-time · Entry level',
       '10,001+ employees · Tobacco Manufacturing',
       'On-site\n\n \nFull-time', 'Remote\n\n \nFull-time',
       'Full-time\n\n \nMid-Senior level', 'On-site\n\n \nPart-time',
       'Hybrid\n\n \nFull-time', 'Remote\n\n \nInternship',
       'On-site\n\n \nInternship', 'Remote\n\n \nContract',
       'Remote\n\n \nPart-time', 'Full-time\n\n \nEntry level',
       

In [1930]:
On_site_remote_options=['On-site','Remote','Hybrid']
On_site_remote_options
# len(On_site_remote_options)

['On-site', 'Remote', 'Hybrid']

In [1931]:
filtered_df = df[df['Job Type'].apply(lambda job_type: any(keyword in job_type for keyword in On_site_remote_options))]
filtered_df
len(filtered_df)

7996

In [1932]:
len(df)-len(filtered_df)

3741

In [1933]:
Job_Type_list=[]
for i in range(len(df['Job Type'].iloc[:])):
    job_type=None
    flag=0
    for keyword in On_site_remote_options:
        if keyword in df['Job Type'].iloc[i]:
            job_type=keyword
            flag=1
            # break
        else:
            pass
            # flag=0
    if flag==1:
        Job_Type_list.append(job_type)
    else:
        Job_Type_list.append('Not specified')

Job_Type_list
len(Job_Type_list)

11737

In [1934]:
len(Job_Type_list)

11737

In [1935]:
set(Job_Type_list)

{'Hybrid', 'Not specified', 'On-site', 'Remote'}

### for verification

In [1936]:
Job_Type_list.count('Not specified')

3741

In [1937]:
len(df)-Job_Type_list.count('Not specified') 
len(df)-Job_Type_list.count('Not specified')==len(df)-Job_Type_list.count('Not specified')

True

In [1938]:
df[df['Job Type']=='On-site\n\n \nFull-time'].iloc[0] #'Hybrid\n\n \nInternship\n\n \nMid-Senior level' '$12,000/yr - $15,000/yr · Full-time · Mid-Senior level'

Job Title                              Manager Digital Lending Business
Job Info              Allied Bank Limited · Lahore, Punjab, Pakistan...
Job Type                                        On-site\n\n \nFull-time
Job Link                  https://www.linkedin.com/jobs/view/3734514085
employees                                             10,001+ employees
Job ID                                                       3734514085
company                                            Allied Bank Limited 
location                    Lahore, Punjab, Pakistan  \n4 days ago\n\n 
Remote/Onsite                                             Not specified
n_applicant                                                         249
job posting time                                             4 days ago
Job Type nElements                                                    2
Job-Type                                                      Full-time
Experience Level                                          Not sp

In [1939]:
df.loc[:,'Remote/Onsite-2']=Job_Type_list

In [1940]:
df['Remote/Onsite-2'].unique()

array(['Not specified', 'On-site', 'Hybrid', 'Remote'], dtype=object)

In [1941]:
df['Remote/Onsite-2'].value_counts()

Remote/Onsite-2
On-site          5720
Not specified    3741
Remote           1725
Hybrid            551
Name: count, dtype: int64

#### how many are specified in Remote/Onsite-2

In [1942]:
df['Remote/Onsite-2'].value_counts()['Not specified']
len(df)-df['Remote/Onsite-2'].value_counts()['Not specified']

7996

#### how many are specified in Remote/Onsite

In [1943]:
df['Remote/Onsite'].value_counts()

Remote/Onsite
Not specified    8396
On-site          2616
Remote            525
Hybrid            200
Name: count, dtype: int64

In [1944]:
len(df)-df['Remote/Onsite'].value_counts()['Not specified']

3341

#### total Remote/Onsite-2 specified jobs

In [1945]:
len(df)-df['Remote/Onsite-2'].value_counts()['Not specified']+len(df)-df['Remote/Onsite'].value_counts()['Not specified']
# 7996+3341

11337

In [1946]:
len(df)-11337

400

***So there are 400 jobs having missing or Not specified Remote/Onsite values***
***

### merging Remote/Onsite and Remote/Onsite-2

***let us investigate pattern***

In [1947]:
df[df['Remote/Onsite']=='Not specified']

Job Title  \
87                         Customer Support Specialist   
94                                Social Media Manager   
96                            Senior Software Engineer   
111                                       Process Lead   
185                      Junior Officer Agri Purchases   
...                                                ...   
15869                                  Unity Developer   
15870  Manager – Growth Chat & SMS (Digital Marketing)   
15872                  Information Security Specialist   
15873        FUNCTIONAL CONSULTANT ORACLE FUSION - HCM   
15874                              Community Associate   

                                                Job Info  \
87     Contour Software · Lahore, Punjab, Pakistan  \...   
94     siyaahat.pk · Lahore District, Punjab, Pakista...   
96     Afiniti · Islamabad, Islāmābād, Pakistan Repos...   
111    BAT · Jhelum, Punjab, Pakistan  \n2 weeks ago\...   
185    Ingredion Incorporated · Faisalabad, Punjab, P...   
...                                                  ...   
15869  Codiea.io · Bahawalpur, Punjab, Pakistan  \n1 ...   
15870  Bank Alfalah Limited · Karāchi, Sindh, Pakista...   
15872  DPL · Islamabad, Islāmābād, Pakistan  \n1 week...   
15873  Level 3 BOS · Karachi Division, Sindh, Pakista...   
15874  The Hive · Karachi Division, Sindh, Pakistan  ...   

                                             Job Type  \
87                            Full-time · Entry level   
94                                          Full-time   
96                       Full-time · Mid-Senior level   
111                      Full-time · Mid-Senior level   
185                           Full-time · Entry level   
...                                               ...   
15869                         On-site\n\n \nFull-time   
15870  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15872       On-site\n\n \nFull-time\n\n \nEntry level   
15873  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15874                         On-site\n\n \nFull-time   

                                            Job Link  \
87     https://www.linkedin.com/jobs/view/3723386608   
94      https://www.linkedin.com/jobs/view/525119752   
96     https://www.linkedin.com/jobs/view/3691553817   
111    https://www.linkedin.com/jobs/view/3728073727   
185    https://www.linkedin.com/jobs/view/3728188227   
...                                              ...   
15869  https://www.linkedin.com/jobs/view/3767951886   
15870  https://www.linkedin.com/jobs/view/3767355025   
15872  https://www.linkedin.com/jobs/view/3772530977   
15873  https://www.linkedin.com/jobs/view/3771943212   
15874  https://www.linkedin.com/jobs/view/3767944844   

                                               employees      Job ID  \
87     1,001-5,000 employees · IT Services and IT Con...  3723386608   
94                                        1-10 employees   525119752   
96          1,001-5,000 employees · Software Development  3691553817   
111            10,001+ employees · Tobacco Manufacturing  3728073727   
185       10,001+ employees · Food and Beverage Services  3728188227   
...                                                  ...         ...   
15869                                     1-10 employees  3767951886   
15870                   5,001-10,000 employees · Banking  3767355025   
15872   51-200 employees · IT Services and IT Consulting  3772530977   
15873  201-500 employees · Outsourcing and Offshoring...  3771943212   
15874                                    11-50 employees  3767944844   

                       company  \
87           Contour Software    
94                siyaahat.pk    
96                    Afiniti    
111                       BAT    
185    Ingredion Incorporated    
...                        ...   
15869               Codiea.io    
15870    Bank Alfalah Limited    
15872                     DPL    
15873             Level 3 BOS    
15874                The Hive 

In [1948]:
df[df['Remote/Onsite-2'].isin(On_site_remote_options)]

C:\Users\BILAL.COM\AppData\Local\Programs\Python\Python311\Lib\site-packages\IPython\core\displayhook.py:281: UserWarning: Output cache limit (currently 1000 entries) hit.
Flushing oldest 200 entries.
  warn('Output cache limit (currently {sz} entries) hit.\n'


Job Title  \
3997                    Team Lead Performance Management   
3998                                 Solutions Architect   
3999                      Customer Support Executive UAE   
4000   Cash Management Product Manager (Mashreq Globa...   
4001                            Senior Territory Manager   
...                                                  ...   
15869                                    Unity Developer   
15870    Manager – Growth Chat & SMS (Digital Marketing)   
15872                    Information Security Specialist   
15873          FUNCTIONAL CONSULTANT ORACLE FUSION - HCM   
15874                                Community Associate   

                                                Job Info  \
3997   Telenor Microfinance Bank Limited · Islamabad,...   
3998   SwipBox · Islāmābād, Pakistan  \n4 days ago\n\...   
3999   Abacus Outsourcing · Lahore, Punjab, Pakistan ...   
4000   Mashreq · Karachi Division, Sindh, Pakistan  \...   
4001   foodpanda · Islamabad, Islāmābād, Pakistan  \n...   
...                                                  ...   
15869  Codiea.io · Bahawalpur, Punjab, Pakistan  \n1 ...   
15870  Bank Alfalah Limited · Karāchi, Sindh, Pakista...   
15872  DPL · Islamabad, Islāmābād, Pakistan  \n1 week...   
15873  Level 3 BOS · Karachi Division, Sindh, Pakista...   
15874  The Hive · Karachi Division, Sindh, Pakistan  ...   

                                             Job Type  \
3997   On-site\n\n \nFull-time\n\n \nMid-Senior level   
3998   On-site\n\n \nFull-time\n\n \nMid-Senior level   
3999   On-site\n\n \nFull-time\n\n \nMid-Senior level   
4000   On-site\n\n \nFull-time\n\n \nMid-Senior level   
4001   On-site\n\n \nFull-time\n\n \nMid-Senior level   
...                                               ...   
15869                         On-site\n\n \nFull-time   
15870  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15872       On-site\n\n \nFull-time\n\n \nEntry level   
15873  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15874                         On-site\n\n \nFull-time   

                                            Job Link  \
3997   https://www.linkedin.com/jobs/view/3734530164   
3998   https://www.linkedin.com/jobs/view/3735939988   
3999   https://www.linkedin.com/jobs/view/3734589987   
4000   https://www.linkedin.com/jobs/view/3734539790   
4001   https://www.linkedin.com/jobs/view/3734537706   
...                                              ...   
15869  https://www.linkedin.com/jobs/view/3767951886   
15870  https://www.linkedin.com/jobs/view/3767355025   
15872  https://www.linkedin.com/jobs/view/3772530977   
15873  https://www.linkedin.com/jobs/view/3771943212   
15874  https://www.linkedin.com/jobs/view/3767944844   

                                               employees      Job ID  \
3997                     1,001-5,000 employees · Banking  3734530164   
3998   51-200 employees · Transportation, Logistics, ...  3735939988   
3999   1,001-5,000 employees · Outsourcing and Offsho...  3734589987   
4000                     1,001-5,000 employees · Banking  3734539790   
4001   5,001-10,000 employees · Internet Marketplace ...  3734537706   
...                                                  ...         ...   
15869                                     1-10 employees  3767951886   
15870                   5,001-10,000 employees · Banking  3767355025   
15872   51-200 employees · IT Services and IT Consulting  3772530977   
15873  201-500 employees · Outsourcing and Offshoring...  3771943212   
15874                                    11-50 employees  3767944844   

                                  company  \
3997   Telenor Microfinance Bank Limited    
3998                             SwipBox    
3999                  Abacus Outsourcing    
4000                             Mashreq    
4001                           foodpanda    
...                                   ...   
15869                          Codiea.io    
15870               Ba

In [1949]:
df[df['Remote/Onsite']=='Not specified'][df['Remote/Onsite-2'].isin(On_site_remote_options)]

C:\Users\BILAL.COM\AppData\Local\Temp\ipykernel_10124\476168604.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df['Remote/Onsite']=='Not specified'][df['Remote/Onsite-2'].isin(On_site_remote_options)]


Job Title  \
3997                    Team Lead Performance Management   
3998                                 Solutions Architect   
3999                      Customer Support Executive UAE   
4000   Cash Management Product Manager (Mashreq Globa...   
4001                            Senior Territory Manager   
...                                                  ...   
15869                                    Unity Developer   
15870    Manager – Growth Chat & SMS (Digital Marketing)   
15872                    Information Security Specialist   
15873          FUNCTIONAL CONSULTANT ORACLE FUSION - HCM   
15874                                Community Associate   

                                                Job Info  \
3997   Telenor Microfinance Bank Limited · Islamabad,...   
3998   SwipBox · Islāmābād, Pakistan  \n4 days ago\n\...   
3999   Abacus Outsourcing · Lahore, Punjab, Pakistan ...   
4000   Mashreq · Karachi Division, Sindh, Pakistan  \...   
4001   foodpanda · Islamabad, Islāmābād, Pakistan  \n...   
...                                                  ...   
15869  Codiea.io · Bahawalpur, Punjab, Pakistan  \n1 ...   
15870  Bank Alfalah Limited · Karāchi, Sindh, Pakista...   
15872  DPL · Islamabad, Islāmābād, Pakistan  \n1 week...   
15873  Level 3 BOS · Karachi Division, Sindh, Pakista...   
15874  The Hive · Karachi Division, Sindh, Pakistan  ...   

                                             Job Type  \
3997   On-site\n\n \nFull-time\n\n \nMid-Senior level   
3998   On-site\n\n \nFull-time\n\n \nMid-Senior level   
3999   On-site\n\n \nFull-time\n\n \nMid-Senior level   
4000   On-site\n\n \nFull-time\n\n \nMid-Senior level   
4001   On-site\n\n \nFull-time\n\n \nMid-Senior level   
...                                               ...   
15869                         On-site\n\n \nFull-time   
15870  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15872       On-site\n\n \nFull-time\n\n \nEntry level   
15873  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15874                         On-site\n\n \nFull-time   

                                            Job Link  \
3997   https://www.linkedin.com/jobs/view/3734530164   
3998   https://www.linkedin.com/jobs/view/3735939988   
3999   https://www.linkedin.com/jobs/view/3734589987   
4000   https://www.linkedin.com/jobs/view/3734539790   
4001   https://www.linkedin.com/jobs/view/3734537706   
...                                              ...   
15869  https://www.linkedin.com/jobs/view/3767951886   
15870  https://www.linkedin.com/jobs/view/3767355025   
15872  https://www.linkedin.com/jobs/view/3772530977   
15873  https://www.linkedin.com/jobs/view/3771943212   
15874  https://www.linkedin.com/jobs/view/3767944844   

                                               employees      Job ID  \
3997                     1,001-5,000 employees · Banking  3734530164   
3998   51-200 employees · Transportation, Logistics, ...  3735939988   
3999   1,001-5,000 employees · Outsourcing and Offsho...  3734589987   
4000                     1,001-5,000 employees · Banking  3734539790   
4001   5,001-10,000 employees · Internet Marketplace ...  3734537706   
...                                                  ...         ...   
15869                                     1-10 employees  3767951886   
15870                   5,001-10,000 employees · Banking  3767355025   
15872   51-200 employees · IT Services and IT Consulting  3772530977   
15873  201-500 employees · Outsourcing and Offshoring...  3771943212   
15874                                    11-50 employees  3767944844   

                                  company  \
3997   Telenor Microfinance Bank Limited    
3998                             SwipBox    
3999                  Abacus Outsourcing    
4000                             Mashreq    
4001                           foodpanda    
...                                   ...   
15869                          Codiea.io    
15870               Ba

In [1950]:
df[df['Remote/Onsite']=='Not specified'][df['Remote/Onsite-2'].isin(On_site_remote_options)]['Remote/Onsite']

C:\Users\BILAL.COM\AppData\Local\Temp\ipykernel_10124\1327332202.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df['Remote/Onsite']=='Not specified'][df['Remote/Onsite-2'].isin(On_site_remote_options)]['Remote/Onsite']


3997     Not specified
3998     Not specified
3999     Not specified
4000     Not specified
4001     Not specified
             ...      
15869    Not specified
15870    Not specified
15872    Not specified
15873    Not specified
15874    Not specified
Name: Remote/Onsite, Length: 7996, dtype: object

In [1951]:
df[df['Remote/Onsite']=='Not specified'][df['Remote/Onsite-2'].isin(On_site_remote_options)]['Remote/Onsite']=df[df['Remote/Onsite-2'].isin(On_site_remote_options)]['Remote/Onsite-2']

C:\Users\BILAL.COM\AppData\Local\Temp\ipykernel_10124\3732851105.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df['Remote/Onsite']=='Not specified'][df['Remote/Onsite-2'].isin(On_site_remote_options)]['Remote/Onsite']=df[df['Remote/Onsite-2'].isin(On_site_remote_options)]['Remote/Onsite-2']


In [1952]:
df['Remote/Onsite'].value_counts()

Remote/Onsite
Not specified    8396
On-site          2616
Remote            525
Hybrid            200
Name: count, dtype: int64

***not changed***

In [1953]:
df['Remote/Onsite'].iloc[0]='remote'
df.loc[0, 'Remote/Onsite'] = 'Remote'

C:\Users\BILAL.COM\AppData\Local\Temp\ipykernel_10124\1029757084.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Remote/Onsite'].iloc[0]='remote'


In [1954]:
df['Remote/Onsite'].iloc[0]

'Remote'

#### replace method
not working

#### Using numpy.where:
working

In [1955]:
# import numpy as np

# condition = (df['Remote/Onsite'] == 'Not specified') & (df['Remote/Onsite-2'].isin(On_site_remote_options))
# df['Remote/Onsite'] = np.where(condition, df['Remote/Onsite-2'], df['Remote/Onsite'])
# ### working

#### Using mask method:
working

In [1956]:
# condition = (df['Remote/Onsite'] == 'Not specified') & (df['Remote/Onsite-2'].isin(On_site_remote_options))
# df['Remote/Onsite'] = df['Remote/Onsite'].mask(condition, df['Remote/Onsite-2'])


#### Using apply method:
working

In [1957]:
# def replace_remote(row):
#     if row['Remote/Onsite'] == 'Not specified' and row['Remote/Onsite-2'] in On_site_remote_options:
#         return row['Remote/Onsite-2']
#     else:
#         return row['Remote/Onsite']

# df['Remote/Onsite'] = df.apply(replace_remote, axis=1)


#### boolean indexing
working

In [1958]:
# condition = (df['Remote/Onsite'] == 'Not specified') & (df['Remote/Onsite-2'].isin(On_site_remote_options))
# df.loc[condition, 'Remote/Onsite'] = df.loc[condition, 'Remote/Onsite-2']


#### as 3997th index to onward pattern changed :slicing

In [1959]:
# df['Remote/Onsite'].iloc[3438:]
# df['Remote/Onsite'].iloc[3438:]=df['Remote/Onsite-2'].iloc[3438:]
# df.loc[3438:, 'Remote/Onsite'] = df['Remote/Onsite-2'].iloc[3438:]  #loc point index so index is 3997
# 3997-3438 more rows
df.loc[3997:, 'Remote/Onsite'] = df['Remote/Onsite-2'].iloc[3438:]

In [1961]:
df['Remote/Onsite'].value_counts()

Remote/Onsite
On-site          8336
Remote           2250
Hybrid            751
Not specified     400
Name: count, dtype: int64

In [1962]:
df.drop(columns=['Remote/Onsite-2'], inplace=True)

#### working on employees feature , Extracting no of employees and industry feature from employees feature

In [109]:
df['employees'].unique()[:10]

array(['11-50 employees', '51-200 employees',
       'Mobeen Naqvi is hiring for this job',
       '1,001-5,000 employees · Software Development',
       '501-1,000 employees · IT Services and IT Consulting',
       '501-1,000 employees · Telecommunications',
       '1,001-5,000 employees · Banking', '201-500 employees',
       '10,001+ employees', '51-200 employees · Human Resources Services'],
      dtype=object)

In [110]:
inconsistent_values_in_employees_feature = ['is hiring for this job', 
                                            'See how you compare', 
                                            'Job poster joined LinkedIn',
                                           'school alum works here',
                                            'alumni work here',
                                            'See recent hiring trends'
                                           ]

In [111]:
L1=[] #for employees 
L2=[] #for industry
for feature_element in df['employees'].iloc[:]:
    flag=0
    for value in inconsistent_values_in_employees_feature:
        if value in feature_element:
            flag=1
            
    if flag==1:
        L1.append('Not specified')
        L2.append('Not specified')
    else:
        if len(feature_element.split('·'))==1:
            L1.append(feature_element)
            L2.append('Not specified')
        elif len(feature_element.split('·'))==2:
            L1.append(feature_element.split('·')[0])
            L2.append(feature_element.split('·')[1])
        else:
            L1.append('Other')
            L2.append('Other')

In [115]:
len(L1)

11737

In [116]:
len(L2)

11737

In [117]:
df.loc[:,'Comp_N_Employees']=L1

In [118]:
df.loc[:,'Comp_Industry']=L2

In [119]:
df['Comp_N_Employees'].unique()

array(['11-50 employees', '51-200 employees', 'Not specified',
       '1,001-5,000 employees ', '501-1,000 employees ',
       '201-500 employees', '10,001+ employees', '51-200 employees ',
       '501-1,000 employees', '1,001-5,000 employees', '1-10 employees ',
       '10,001+ employees ', '1-10 employees', '5,001-10,000 employees',
       '201-500 employees ', '5,001-10,000 employees ',
       '11-50 employees ', '1 company alum works here',
       'Industrial Automation'], dtype=object)

In [120]:
len(df['Comp_N_Employees'].unique())

19

In [121]:
df['Comp_N_Employees'].value_counts()

Comp_N_Employees
11-50 employees              2258
5,001-10,000 employees       1481
51-200 employees             1407
1,001-5,000 employees        1228
10,001+ employees            1133
Not specified                 946
1-10 employees                674
201-500 employees             596
51-200 employees              464
201-500 employees             346
501-1,000 employees           340
11-50 employees               273
501-1,000 employees           189
1,001-5,000 employees         165
1-10 employees                126
5,001-10,000 employees         60
10,001+ employees              49
1 company alum works here       1
Industrial Automation           1
Name: count, dtype: int64

In [122]:
df['Comp_Industry'].unique()

array(['Not specified', ' Software Development',
       ' IT Services and IT Consulting', ' Telecommunications',
       ' Banking', ' Human Resources Services',
       ' Technology, Information and Internet',
       ' Retail Apparel and Fashion', ' International Affairs',
       ' Higher Education', ' Information Technology & Services',
       ' Financial Services', ' Packaging and Containers Manufacturing',
       ' Executive Offices', ' Food and Beverage Manufacturing',
       ' Food and Beverage Services', ' Tobacco Manufacturing',
       ' Business Consulting and Services', ' Entertainment Providers',
       ' Wholesale Building Materials',
       ' Outsourcing and Offshoring Consulting',
       ' Pharmaceutical Manufacturing', ' Motor Vehicle Manufacturing',
       ' Non-profit Organizations', ' Research Services',
       ' Advertising Services', ' Beverage Manufacturing',
       ' Outsourcing/Offshoring', ' Hospitals and Health Care',
       ' Farming', ' Oil and Gas', ' Real Est

In [123]:
len(df['Comp_Industry'].unique())

112

In [124]:
df['Comp_Industry'].value_counts()

Comp_Industry
Not specified                                6346
 IT Services and IT Consulting               1758
 Software Development                         612
 Telecommunications                           310
 Real Estate                                  290
                                             ... 
 Renewable Energy Equipment Manufacturing       1
 Writing and Editing                            1
 Furniture                                      1
 Research                                       1
 Retail Luxury Goods and Jewelry                1
Name: count, Length: 112, dtype: int64

In [1963]:
df.columns

Index(['Job Title', 'Job Info', 'Job Type', 'Job Link', 'employees', 'Job ID',
       'company', 'location', 'Remote/Onsite', 'n_applicant',
       'job posting time', 'Job Type nElements', 'Job-Type',
       'Experience Level'],
      dtype='object')

In [127]:
len(df.columns)

14

## End Note

***We have to work on location feature to make it meaningfull***

***n_applicant_have_4686_null_values,investigate it***

***make a new ratio feature of posting time and n_applicant feature***

****

# To be noted that:
***remove dulicates first and then remove null values to avoid warning***

***first remove duplicates and then remove null values because this way we can avoid warning***
